### ISSUE TITLE
cducrest-brainbot - Attacker can abuse FeeManager's collectMintFee to drain FeeManager

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-04-titles-judging#139]
cducrest-brainbot

high

# Attacker can abuse FeeManager's collectMintFee to drain FeeManager

## Summary

The function to collect fee on FeeManager `collectMintFee()` is unprotected and can be called by anyone (especially the fee recipient of a specific work) to drain the ether in `FeeManager`.

## Vulnerability Detail

The function `collectMintFee()` has no authorization and can be called by anyone:

```solidity
    function collectMintFee(
        IEdition edition,
        uint256 tokenId_,
        uint256 amount_,
        address payer_,
        address referrer_,
        Strategy calldata strategy_
    ) external payable {  // @audit-issue no protection, can siphon any funds
        _collectMintFee(
            edition, tokenId_, amount_, payer_, referrer_, getMintFee(strategy_, amount_)
        );
    }
```

The internal function `_collectMintFee()` will compute the fee to be paid to edition referrer, mint referrer (provided input), (edition, tokenId) fee receivers, and protocol. It will then `route` (transfer) the fee:

```solidity
    function _collectMintFee(
        IEdition edition_,
        uint256 tokenId_,
        uint256 amount_,
        address payer_,
        address referrer_,
        Fee memory fee_
    ) internal {
        if (fee_.amount == 0) return;

        ...

        uint256 protocolFee = protocolFlatFee * amount_;
        uint256 protocolShare;
        if (fee_.amount == protocolFee) {
            protocolShare = protocolFee * protocolFeeshareBps / MAX_BPS;
        } else {
            protocolShare = protocolFee;
        }

        _route(
            Fee({asset: fee_.asset, amount: fee_.amount - protocolShare}),
            _feeReceivers[getRouteId(edition_, tokenId_)],
            payer_
        );

        uint256 referrerShare =
            _splitProtocolFee(edition_, fee_.asset, protocolShare, payer_, referrer_);
        emit FeeCollected(address(edition_), tokenId_, fee_.asset, fee_.amount, referrerShare);
    }
```

This can be called by anyone to transfer any amount of eth from `FeeManager` to a split of addresses.

## Impact

Anyone can create a new edition through `TitlesCore` which will create a route on the `FeeManager` with user-provided inputs. This will also set up the edition's referrer with user-provided input. Lastly, the mint referrer is provided as input of the `FeeManager.collectMintFee()` function.

Anyone can control 3 out of the 4 addresses receiving ether as a result of `collectMintFee()`, the uncotrolled address being `protocolFeeReceiver` which receives a limited amount of eth from the mint.

As a result, anyone can drain all the ether from `FeeManager` minus some small protocol fee amount.

The documentation for the contest states about the `ADMIN_ROLE` and the `withdraw()` function of the `FeeManager`:

> ADMIN_ROLE (Trusted) => Granted by the deployer to internal, trusted addresses only.
>
> On FeeManager, this role can:
> 3) Withdraw any funds locked in the contract (withdraw). This is the only way to withdraw funds from the contract.

Which does not hold.

This vulnerability can additionally be exploited by any fee recipient of a mint to drain any overpaid amount to the FeeManager as a re-entrancy upon receiving fees. Knowing the Edition transfer the total `msg.value` received during a mint to the `FeeManager`, any fee recipient may drain the overpaid user ether that will never be reimbursed to the user.

## Code Snippet

https://github.com/sherlock-audit/2024-04-titles/blob/d7f60952df22da00b772db5d3a8272a988546089/wallflower-contract-v2/src/fees/FeeManager.sol#L183-L213

## Tool used

Manual Review

## Recommendation

Do not hold funds in FeeManager. Currently the FeeManager receives the total `msg.value` from the Edition during a mint, which is problematic as this may leave ether on the FeeManager that can be drained by anyone.

Duplicate of #269

**Comment by cducrest** [comment]
Escalate

This is a valid high for two reasons.

1. The behaviour of the contract differs from what is expected from the readme of the contest as explained in the issue. Readme states that only `ADMIN_ROLE` should be able to withdraw funds from the FeeManager. I prove that other parties can siphon funds from FeeManager. Contest's readme supersedes Sherlock's judging rules per Sherlock's Hierarchy of truth.

2. The issue states that siphoning funds from FeeManager can be used in a re-entrancy attack by any fee recipient to drain overpaid amount by the user to the FeeManager through Edition during minting. This means loss of any overpaid user ETH. This is an analogue impact to what is described in https://github.com/sherlock-audit/2024-04-titles-judging/issues/268.

The difference in between this issue and https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 is that https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 assumes that `Edition` does not send the total `msg.value` to `FeeManager` during mint which is not true. As such, https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 explains that excess ETH of the victim can be stolen from `Edition` (which is not true in the current code version) while this issue explains that excess ETH of the victim can be stolen from `FeeManager` .

A point of view could be that this issue and https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 are close enough to be duplicate. I am not sure what judges think.

Perhaps this issue should have been split into two different issues as there are actually two impacts even though the attack vector/remedy is similar.

**Comment by sherlock-admin3** [comment]
> Escalate
> 
> This is a valid high for two reasons.
> 
> 1. The behaviour of the contract differs from what is expected from the readme of the contest as explained in the issue. Readme states that only `ADMIN_ROLE` should be able to withdraw funds from the FeeManager. I prove that other parties can siphon funds from FeeManager. Contest's readme supersedes Sherlock's judging rules per Sherlock's Hierarchy of truth.
> 
> 2. The issue states that siphoning funds from FeeManager can be used in a re-entrancy attack by any fee recipient to drain overpaid amount by the user to the FeeManager through Edition during minting. This means loss of any overpaid user ETH. This is an analogue impact to what is described in https://github.com/sherlock-audit/2024-04-titles-judging/issues/268.
> 
> The difference in between this issue and https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 is that https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 assumes that `Edition` does not send the total `msg.value` to `FeeManager` during mint which is not true. As such, https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 explains that excess ETH of the victim can be stolen from `Edition` (which is not true in the current code version) while this issue explains that excess ETH of the victim can be stolen from `FeeManager` .
> 
> A point of view could be that this issue and https://github.com/sherlock-audit/2024-04-titles-judging/issues/268 are close enough to be duplicate. I am not sure what judges think.
> 
> Perhaps this issue should have been split into two different issues as there are actually two impacts even though the attack vector/remedy is similar.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by cvetanovv** [comment]
I agree with the escalation, and this issue is a duplicate of #266.

**Comment by Evert0x** [comment]
Result:
High
Duplicate of #266 

---

actual severity still depends on 266

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [cducrest](https://github.com/sherlock-audit/2024-04-titles-judging/issues/139/#issuecomment-2107505604): accepted

**Comment by WangSecurity** [comment]
This report is not a duplicate of #266 because 266 talks about ERC20 tokens which are not used in the current version of the protocol. Actually, it's a dup of #425 and will be duplicated accordingly.

**Comment by cducrest** [comment]
While this issue (https://github.com/sherlock-audit/2024-04-titles-judging/issues/139), https://github.com/sherlock-audit/2024-04-titles-judging/issues/383, https://github.com/sherlock-audit/2024-04-titles-judging/issues/357, and https://github.com/sherlock-audit/2024-04-titles-judging/issues/425 all point to the fact that `collectMintFee()` can be exploited to drain funds from the `FeeManager`, only this issue (https://github.com/sherlock-audit/2024-04-titles-judging/issues/139) explain that this can be used in a re-entrancy attack by any legit fee recipient of a mint to steal the overpaid amount by users.

The difference in impact is that in other issues the loss of funds only appear under the specific scenario where the `FeeManager` holds funds for an extended period of time (across multiple transactions) which is unlikely given the behaviour of `FeeManager` and flow of funds during fee payment. While this issue (https://github.com/sherlock-audit/2024-04-titles-judging/issues/139) shows a guaranteed loss of funds for any user providing excess eth for a mint.

Based on these two judges decisions: https://github.com/sherlock-audit/2023-02-hats-judging/issues/41#issuecomment-1494727841, and https://github.com/sherlock-audit/2023-03-teller-judging/issues/26 where the same flaw (same lines of codes / function pointed out) resulted in two different issues with different impact, I claim that only this issue (https://github.com/sherlock-audit/2024-04-titles-judging/issues/139) has High impact and other issues while still valid should have Medium impact.

**Comment by WangSecurity** [comment]
Firstly, historical decisions are not sources of truth. Secondly, I agree with your words, indeed FeeManager can only hold funds if there was a direct transfer into the contract (user mistake) or due to issue in #269 (excess ETH during minting). This report identifies the problem of excess ETH during minting and shows how it can be stolen. Therefore, I believe it should be a duplicate of #269, cause this issue only possible due to the vulnerability in #269 and if it's fixed, then this issue would be invalid due to user mistake.

Hence, planning to accept the escalation and duplicate with #269 as it shows an additional impact of the same core issue.

**Comment by cducrest** [comment]
I disagree with this point of view.

The Readme of the contest states explicitly:

> Are there any protocol roles? Please list them and provide whether they are TRUSTED or RESTRICTED, or provide a more comprehensive description of what a role can and can't do/impact.
> 
> ADMIN_ROLE (Trusted) => Granted by the deployer to internal, trusted addresses only.
>
>   On FeeManager, this role can:
>   3) Withdraw any funds locked in the contract (withdraw). This is the only way to withdraw funds from the contract.

It also states:

> For permissioned functions, please list all checks and requirements that will be made before calling the function.
> 
> FeeManager:
> - withdraw => checks that the caller is the owner or has the ADMIN_ROLE

Meaning only the admin should be able to withdraw locked funds in the FeeManager. As proved by this issue and all aforementioned duplicates, these requirements do not hold and all the duplicates including this issue should be valid as "Contest's readme supersedes Sherlock's judging rules per Sherlock's Hierarchy of truth."

These arguments were already provided during the escalation on this issue.

My point of view is that this issue provides an **additional** attack vector stealing excess eth provided by the user via a re-entrancy attack which is a high vulnerability. You are correct that if https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 is fixed then this **additional** attack vector does not hold. In any case, the discrepancy in between the Readme's description of the admin role and permission to withdraw from the FeeManager still holds.

I don't think searcher can make assumptions about how the code will evolve and whether https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 will be fixed by (1) not sending excess eth to FeeManager or (2) reimbursing the final balance of FeeManager back to the user. In case (2) the described re-entrancy issue would still hold. This is just an example to justify that these issues should be reported, hence awarded.

**Comment by WangSecurity** [comment]
Firstly, I didn't argue that it allows not admins get funds from the FeeManager, the problem is how funds get into the FeeManager. There are only two ways there can be excess funds in FeeManager that I expressed in the previous message. As we both agree this is an additional attack vector to #269 and if #269 is fixed, then this issue it fixed as well. 

I agree with you that the fee receiver can still re-enter in the FeeManager contract even if the #269 is fixed, but I believe there is no impact or funds stolen (assuming transferring funds directly is the user mistake and won't happen). Therefore, I believe this issue identifies core issue as #269 that excess `msg.value` can be sent in FeeManager, but identifies another impact to this. Still I believe it should be medium based on the comment left under #269. Hence, decision remains the same, duplicate this report with #269 .

**Comment by cducrest** [comment]
I understand the words and sentences you use.

I believe if https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 is fixed and not this issue, a problem remains in the code base.

The first problem that definitely remains is that non-admin users can withdraw funds from the FeeManager. This is problem for the protocol team as they state the expectations they have for the code in the Readme of the contest. The protocol team intends for the FeeManager to eventually hold funds otherwise what is this doing: https://github.com/sherlock-audit/2024-04-titles/blob/d7f60952df22da00b772db5d3a8272a988546089/wallflower-contract-v2/src/fees/FeeManager.sol#L352-L357 ?

Secondly, depending on how protocol decides to fix https://github.com/sherlock-audit/2024-04-titles-judging/issues/269, this vulnerability could remain open. It luckily happens that the fix proposed by the researcher in https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 also prevents this vulnerability. 

If the selected fix would have been to refund the fees from the FeeManager back to the user or to the Edition, then this vulnerability would have remained open. That is why I believe this issue is different from https://github.com/sherlock-audit/2024-04-titles-judging/issues/269, is a High vulnerability issue and valid.

If you take for example https://github.com/sherlock-audit/2024-04-titles-judging/issues/151, the researcher suggests using `_refundExcess()` in `FeeManager` at the end of `FeeManager.collectMintFee()`. This would mean that during the mint, the overpaid fees is held by `FeeManager` and the re-entrancy attack stealing the funds of the `FeeManager` can be performed. This will lower the balance of `FeeManager` and less value (or no value) will be refunded to the overpaying user which is a guaranteed loss of funds to the user. 

Here is a list of issues duplicate with https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 that recommend a fix that would leave this vulnerability open: https://github.com/sherlock-audit/2024-04-titles-judging/issues/392, https://github.com/sherlock-audit/2024-04-titles-judging/issues/343, https://github.com/sherlock-audit/2024-04-titles-judging/issues/163, https://github.com/sherlock-audit/2024-04-titles-judging/issues/380, https://github.com/sherlock-audit/2024-04-titles-judging/issues/211, https://github.com/sherlock-audit/2024-04-titles-judging/issues/76, https://github.com/sherlock-audit/2024-04-titles-judging/issues/331, https://github.com/sherlock-audit/2024-04-titles-judging/issues/351, https://github.com/sherlock-audit/2024-04-titles-judging/issues/151, https://github.com/sherlock-audit/2024-04-titles-judging/issues/164, https://github.com/sherlock-audit/2024-04-titles-judging/issues/3, https://github.com/sherlock-audit/2024-04-titles-judging/issues/190, https://github.com/sherlock-audit/2024-04-titles-judging/issues/80, https://github.com/sherlock-audit/2024-04-titles-judging/issues/184, https://github.com/sherlock-audit/2024-04-titles-judging/issues/186, https://github.com/sherlock-audit/2024-04-titles-judging/issues/399.

Here is a list of issues duplicate with https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 that recommend a fix that would potentially leave this vulnearbility open: https://github.com/sherlock-audit/2024-04-titles-judging/issues/419, https://github.com/sherlock-audit/2024-04-titles-judging/issues/205, https://github.com/sherlock-audit/2024-04-titles-judging/issues/318, https://github.com/sherlock-audit/2024-04-titles-judging/issues/333, https://github.com/sherlock-audit/2024-04-titles-judging/issues/159.

In conclusion, if by any chance a different fix than the one propose by https://github.com/sherlock-audit/2024-04-titles-judging/issues/269 would be implemented (e.g. a fix from the aforementioned issues), the re-entrancy vulnerability would remain open and the protocol would be vulnerable. This submission should be rewarded in my opinion.

Thank you for your time and understanding.

**Comment by WangSecurity** [comment]
I see and agree that other potential fixes wouldn't mitigate this problem, but we both agree that the fix of #269 will fix this issue as well. I understand there there is still potential for other users to withdraw funds. But the only funds in that case are from direct transfers, then it's user mistake. The existence of `withdraw` function itself doesn't mean direct transfers are not considered a user mistake. If there are any business flows seen from the code leading to funds just sit in the contract, then it would be a valid argument. But from what I know, there are no specific business flows in the current implementation that require it. Hence, I assume the only way for ETH to be in the FeeManager is user mistake (if #269 is fixed).

That is why I believe it should be the duplicate with #269 plus reasons explained in my above messages.
